### PR TITLE
Remove AuthProvider / MetaProvider in favor of redux

### PR DESCRIPTION
**Pull request by jgradim**
~~Note: depends on https://github.com/lifeonmarspt/mentorados-backend/pull/7 for https://github.com/lifeonmarspt/mentorados-frontend/commit/d0647fc0c98ee82d56ddb66798eead377a310ac8 to work properly~~ (merged)

### CODE REVIEW

File: app/actions/mentors.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+const setMentors = createAction(SET_MENTORS);
+
+export const getMentors = (filters = {}) => {
+  return dispatch => {
+    return mentors
+      .index(filters)
+      .then(response => {
+        dispatch(setMentors(response.data));
+        return Promise.resolve(response.data);
```
Comment (hugopeixoto) [inline app/actions/mentors.js]: can't you just `return response.data`? (The same applies to all instances of this pattern below)

File: app/actions/mentors.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+      })
+      .catch(error => Promise.reject(error));
```
Comment (hugopeixoto) [inline app/actions/mentors.js]: Does this do anything? isn't the catch the result of a rejected promise?

File: app/components/Layout.jsx [hunk 3]
```diff
@@ -1,18 +1,39 @@
...
 import { Route } from "react-router";
 import { Link } from "react-router-dom";
+import { compose } from "recompose";
+import { connect } from "react-redux";
+import { withRouter } from "react-router-dom";
 
 import Header from "components/Header";
 import Footer from "components/Footer";
 import Toaster from "components/Toaster";
+import Blocked from "components/pages/Blocked";
```
Comment (hugopeixoto) [inline app/components/Layout.jsx]: Move this to `components/Blocked`

File: app/components/Layout.jsx [hunk 4]
```diff
@@ -1,18 +1,39 @@
...
+      getCurrentUser(),
+      getMeta(),
+    ])
+    .finally(() => this.setState({ loading: false }));
   }
 
   render() {
+    const { currentUser } = this.props;
+
+    if (this.state.loading) return <div id="layout">Loading...</div>;
```
Comment (hugopeixoto) [inline app/components/Layout.jsx]: Translate this

File: app/components/elements/Filters/index.js [hunk 5]
```diff
@@ -31,7 +33,9 @@ class Filters extends React.Component {
   }
 
   render() {
-    const { t, filters: { careers, traits } } = this.props;
+    const { t, meta, filters } = this.props;
+
+    if (isEmpty(meta)) return <div>Loading...</div>;
```
Comment (hugopeixoto) [inline app/components/elements/Filters/index.js]: We're already checking for loading in `Layout`

File: app/components/elements/Login/index.js [hunk 6]
```diff
@@ -29,24 +30,22 @@ class Login extends React.Component {
...
-        this.context.session.doLogin(result.data);
-      })
-      .catch((error) => {
-        this.setState({ errors: errorTransform(error, { 404: "Invalid credentials" }) });
-      });
+    const { email, password } = this.state.fields;
+
+    this.props
+    .login(email, password)
+    .catch(error => this.setState({ errors: errorTransform(error, { 404: "Invalid credentials" })}));
```
Comment (hugopeixoto) [inline app/components/elements/Login/index.js]: Translation missing

File: app/components/pages/Account/index.js [hunk 7]
```diff
@@ -161,15 +172,20 @@ class Account extends React.Component {
                 <div className="pure-control-group selection-list">
                   <label>{t("form.career_ids.label")}</label>
```
Comment (hugopeixoto) [inline app/components/pages/Account/index.js:173]: `htmlFor`?

File: app/components/pages/Account/index.js [hunk 8]
```diff
@@ -161,15 +172,20 @@ class Account extends React.Component {
...
                     checked={this.formValue("career_ids")}
-                    onChange={this.onListChange}
+                    onChange={this.onCareersChange}
                   />
                 </div>
 
                 <div className="pure-control-group">
-                  <label html="traits_list">{t("form.traits_list.label")}</label>
-                  <input type="text" name="traits_list" placeholder={t("form.traits_list.placeholder")} value={this.formValue("traits_list").join(",")} onChange={this.onInputChange} />
+                  <label htmlFor="traits_list">{t("form.traits_list.label")}</label>
```
Comment (hugopeixoto) [inline app/components/pages/Account/index.js]: `htmlFor` leads nowhere

File: app/components/pages/Mentors/index.js [hunk 9]
```diff
@@ -22,54 +26,41 @@ const sortByKey = (array, keyFunction) => (
...
-        });
-      });
-  }
-
-  componentWillUnmount() {
-    this.getMentorsPromise.cancel();
-  }
+    this.props
+      .getMentors(filters)
+      .then(() => this.setState({ loading: false }))
```
Comment (hugopeixoto) [inline app/components/pages/Mentors/index.js]: use `finally`

File: app/components/pages/Mentors/index.js [hunk 10]
```diff
@@ -80,16 +71,16 @@ class Mentors extends React.Component {
         <div className="pure-u-1 pure-u-md-18-24 posts">
           <h1 className="content-subhead">
             {t("list.title")}
-            {this.state.loading && <strong> {t("list.loading")}</strong>}
+            {loading && <strong> {t("list.loading")}</strong>}
           </h1>
 
-          {mentors && mentors.length === 0 && (
+          {!loading && isEmpty(mentors) && (
```
Comment (hugopeixoto) [inline app/components/pages/Mentors/index.js]: use length instead

File: app/enhancers/adminRoute.js [hunk 11]
```diff
@@ -1,21 +1,24 @@
 import React from "react";
 import PropTypes from "prop-types";
+import { compose } from "recompose";
+import { connect } from "react-redux";
 
-const adminRoute = (Component) => (
+const adminRoute = (Component) => compose(
+  connect(({ currentUser }) => ({ currentUser })),
+)(
   class AdminRoute extends React.Component {
```
Comment (hugopeixoto) [inline app/enhancers/adminRoute.js]: More toast

File: app/enhancers/anonymousRoute.js [hunk 12]
```diff
@@ -1,22 +1,24 @@
 import React from "react";
 import PropTypes from "prop-types";
+import { connect } from "react-redux";
 
-const anonymousRoute = (Component) => (
+const anonymousRoute = (Component) => connect(
```
Comment (hugopeixoto) [inline app/enhancers/anonymousRoute.js]: Either use compose here or stop using it above

File: app/reducers/session.js [hunk 13]
```diff
@@ -0,0 +1,19 @@
+import {
```
Comment (hugopeixoto) [inline app/reducers/session.js]: baleet me

File: app/resources/mentors.jsx [hunk 14]
```diff
@@ -12,20 +14,18 @@ import { defaultRoutes, defaultActions } from "reactAdmin/helpers";
 
 import { api } from "lib/api";
 
-const createMetaChoiceProvider = (field) => (class extends React.Component {
+const createMetaChoiceProvider = (field) => compose(
+  connect(({ meta }) => ({ meta })),
+)(class extends React.Component {
   static displayName = "createMetaChoiceProvider";
```
Comment (hugopeixoto) [inline app/resources/mentors.jsx]: move this to the class declaration

File: app/components/pages/Account/index.js (hunk 7 above)
Comment (jgradim) [reply app/components/pages/Account/index.js:173]: Doesn't really make sense here, as we have a list of inputs, so clicking the label should not change any of them

File: app/components/Layout.jsx [hunk 15]
```diff
@@ -1,33 +1,69 @@
...
+  }
+
+  getUserAndMeta() {
+    const { getMeta, getCurrentUser } = this.props;
+
+    Promise.all([
+      getCurrentUser(),
+      getMeta(),
+    ])
+    .catch(() => {})
```
Comment (hugopeixoto) [inline app/components/Layout.jsx]: What's up with this catch?

**Comment by jgradim** [comment]
@hugopeixoto updated

**Pull request merged**